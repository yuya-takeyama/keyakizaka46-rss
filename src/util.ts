const baseUrl = 'http://www.keyakizaka46.com';

export const path = (path: string) => baseUrl + path;
