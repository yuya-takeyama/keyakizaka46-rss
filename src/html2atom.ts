import { Feed } from 'feed';
import { JSDOM } from 'jsdom';
import { path } from './util';

const feed = new Feed({
  title: '欅坂46公式サイト',
  description: '「坂道シリーズ」第2弾　欅坂46',
  feed: path('/'),
  id: path('/'),
  link: path('/'),
  feedLinks: {},
  copyright: 'Seed & Flower LLC',
});

export const html2atom = (html: string): string => {
  const dom = new JSDOM(html);
  const elems = dom.window.document.querySelectorAll('.box-news ul li');

  for (const elem of elems) {
    const dateElem = elem.querySelector('.date');
    const date = dateElem && new Date(dateElem.innerHTML);
    const linkElem = elem.querySelector('.text a');
    let link: string | null;
    let title: string | null;
    if (linkElem) {
      link = linkElem.attributes && linkElem.getAttribute('href');
      title = linkElem.innerHTML;

      if (link && date) {
        feed.addItem({
          title,
          link,
          date,
        });
      }
    }
  }

  return feed.atom1();
};
