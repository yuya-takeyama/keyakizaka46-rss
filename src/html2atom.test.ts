import { join } from 'path';
import { readFile as origReadFile } from 'fs';
import { promisify } from 'util';
import { html2atom } from './html2atom';

const readFile = promisify(origReadFile);

const newsFixturesDir = join(__dirname, 'fixtures', 'news');

describe('#html2atom', () => {
  it('returns an Atom feed', async () => {
    const html = (
      await readFile(join(newsFixturesDir, 'news.html'))
    ).toString();
    expect(html2atom(html)).toMatchSnapshot();
  });
});
