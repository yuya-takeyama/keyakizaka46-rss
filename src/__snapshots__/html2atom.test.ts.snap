// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#html2atom returns an Atom feed 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id>https://www.keyakizaka46.com/</id>
    <title>欅坂46公式サイト</title>
    <updated>2020-05-16T19:16:25.238Z</updated>
    <generator>https://github.com/jpmonette/feed</generator>
    <link rel=\\"alternate\\" href=\\"https://www.keyakizaka46.com/\\"/>
    <link rel=\\"self\\" href=\\"https://www.keyakizaka46.com/\\"/>
    <subtitle>「坂道シリーズ」第2弾　欅坂46</subtitle>
    <rights>Seed &amp; Flower LLC</rights>
    <entry>
        <title type=\\"html\\"><![CDATA[5月15日(金)～WEBサイト「billboard JAPAN」の「#stayhome プレイリスト企画」に渡邉理佐と田村保乃が登場!]]></title>
        <id>/s/k46o/news/detail/M01655?ima=0000</id>
        <link href=\\"/s/k46o/news/detail/M01655?ima=0000\\">
        </link>
        <updated>2020-05-14T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[佐藤詩織 一時的な活動参加のお知らせ]]></title>
        <id>/s/k46o/news/detail/O00198?ima=0000</id>
        <link href=\\"/s/k46o/news/detail/O00198?ima=0000\\">
        </link>
        <updated>2020-05-01T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[欅坂46 4th YEAR ANNIVERSARY オフィシャルグッズ  一部アイテムラインナップ公開 &amp; インターネット通販のお知らせ]]></title>
        <id>/s/k46o/news/detail/G00105?ima=0000</id>
        <link href=\\"/s/k46o/news/detail/G00105?ima=0000\\">
        </link>
        <updated>2020-04-05T15:00:00.000Z</updated>
    </entry>
</feed>"
`;
